CHANGELOG.md

```diff
@@ -13,6 +13,7 @@ Changelog
 -   Fixed the QCS access request link in the README (@amyfbrown, gh-1171).
 -   Fix the SDK download link and instructions in the docs (@amyfbrown, gh-1173).
 -   Removed HALT from valid Protoquil / supported Quil. (@kilimanjaro, gh-1176).
+-   Fix spurious PHASE and conditional in exponential_map (@jlapeyre, gh-1182).
```

Comment:
I don't have a good sense for how likely it is that anyone depended on the old behavior. Do you think it's worth calling out in the changelog here that this is a potentially user visible / breaking change?

Likewise, should this go under the "Improvements and Changes" section, or is this more of a bugfix?

Reply:
I think

- Worth calling out in the changelog. In addition to the (admittedly small) chance that users were explicitly relying on the old behavior, they could also be implicitly relying on it: `qc.compiler.native_quil_to_executable(Program())` will fail with an error about an empty program.
- Improvement, I think. The previous `exponential_map`, with the PHASE instructions etc, was correct in that the corresponding quantum operation was the identity.

Reply:
> Improvement

Agreed.

> Worth calling out

That's a good point. There are two issues. The one about the empty program only concerns removing `PHASE`, not the inside-outside the closure one. The more I think about it, the more I think these should not both be changed in the same PR. Each one separately could break code.

Reply:
> The more I think about it, the more I think these should not both be changed in the same PR. Each one separately could break code.

This brings up the (somewhat painful now that we've been through multiple rounds of code review already :D) question of whether these changes are really worth it, since they might break user code, however unlikely.

I don't have a strong opinion either way. I agree it seems pretty unlikely anyone was depending on the old behavior, but you'd be surprised.

I also admit that I don't understand the use cases well enough to understand why the phase thing in the identity case is annoying / undesirable, so I defer to others on that one.

As for the inside-out closure change, I agree the new version is an improvement, but in the absence of any users complaining about it, I'd be inclined to let a sleeping dog lie. I also won't make a stink though if others feel like it's a minor enough breakage that it shouldn't matter... Maybe we should have a label for github issues that introduce breaking changes, to mark them for inclusion on some bright future day when we do a major version bump?

Reply:
> whether these changes are really worth it

Well, this is certainly not urgent, and it's somewhat contentious. The closure thing is just about code quality and clarity, not user experience.

> phase thing in the identity case

AFAICT it was put there for some kind of testing or debugging purpose. It has no business being there in the production code. It is not even documented! You could debate like this though

Pro PHASE:
If the user asked for `exponential_map` of the identity, then we should not second guess the motive. We should return a function that does something non-trivial with it's parameter (phase angle) and still gives the correct behavior. Rotating back and forth is probably the simplest thing you can do. And, I guess do it on qubit zero, because... why put it on a different one ?

Anti PHASE:
`exponential_map` will be called on gates that were generated programmatically, not literally. The user would expect that it does the economic thing, eliding the operation on the identity. And certainly would expect that it does not do something arbitrary like play with the first qubit. Furthermore, if a nontrivial program is returned, It will later be elided by the compiler in any case.